<a href="https://colab.research.google.com/github/rociotarda/TP1---Datos/blob/main/TP1_duplicated_buildings.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from google.colab import files
uploaded = files.upload()

Saving train_values.csv to train_values.csv
Saving train_labels.csv to train_labels.csv


In [3]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import io
  
train_labels = pd.read_csv(io.BytesIO(uploaded['train_labels.csv']))
train_labels.head()

building_id  damage_grade
0       802906             3
1        28830             2
2        94947             3
3       590882             2
4       201944             3

In [4]:
train_values = pd.read_csv(io.BytesIO(uploaded['train_values.csv']))
train_values.head(10)

building_id  ...  has_secondary_use_other
0       802906  ...                        0
1        28830  ...                        0
2        94947  ...                        0
3       590882  ...                        0
4       201944  ...                        0
5       333020  ...                        0
6       728451  ...                        0
7       475515  ...                        0
8       441126  ...                        0
9       989500  ...                        0

[10 rows x 39 columns]

In [5]:
train_values.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 260601 entries, 0 to 260600
Data columns (total 39 columns):
 #   Column                                  Non-Null Count   Dtype 
---  ------                                  --------------   ----- 
 0   building_id                             260601 non-null  int64 
 1   geo_level_1_id                          260601 non-null  int64 
 2   geo_level_2_id                          260601 non-null  int64 
 3   geo_level_3_id                          260601 non-null  int64 
 4   count_floors_pre_eq                     260601 non-null  int64 
 5   age                                     260601 non-null  int64 
 6   area_percentage                         260601 non-null  int64 
 7   height_percentage                       260601 non-null  int64 
 8   land_surface_condition                  260601 non-null  object
 9   foundation_type                         260601 non-null  object
 10  roof_type                               260601 non-null 

**Deberiamos categorizar las cols que veamos correspondientes y así se reduciría bastante el almacenamiento utilizado**

Veamos si hay registros duplicados, sin tener en cuenta el building_id

In [ ]:
train_values[train_values.duplicated(subset=list(train_values.columns)[1:],keep=False)].shape[0]

28544

In [30]:
train_values[train_values.duplicated(subset=list(train_values.columns)[1:],keep=False)].head(10)

building_id  ...  has_secondary_use_other
1          28830  ...                        0
3         590882  ...                        0
12        444381  ...                        0
18        354011  ...                        0
48        987864  ...                        0
59        302199  ...                        0
71        708530  ...                        0
76        425720  ...                        0
103        96470  ...                        0
105       820425  ...                        0

[10 rows x 39 columns]

De un total de 260601 registros, vemos que hay 28544 que son duplicados. Veamos si alguno de ellos es debido a que hay registros duplicados con building_id

In [6]:
len(set(train_values[train_values.duplicated(subset=list(train_values.columns)[1:],keep=False)].building_id))

28544

Tambien lo podemos ver con un value_counts de la columna:

In [7]:
train_values['building_id'].value_counts()

1052670    1
847304     1
368102     1
729986     1
900578     1
          ..
56753      1
946311     1
15789      1
13740      1
524288     1
Name: building_id, Length: 260601, dtype: int64

Esto quiere decir que estos registros "duplicados" comparten todas las columnas, pero parecería ser que no tienen el mismo building_id. Lo que puede suceder es que son edificios "residenciales" que se construyeron en el mismo lugar, con el mismo tamaño y de los mismos materiales de construcción. Veamos por ejemplo la cantidad de registros por geo_level_3_id:

In [9]:
train_values['geo_level_3_id'].value_counts()

633      651
9133     647
621      530
11246    470
2005     466
        ... 
6095       1
3599       1
7886       1
9994       1
11851      1
Name: geo_level_3_id, Length: 11595, dtype: int64

Viendo anteriormente alguno de los registros que parecen duplicados, agarremos el geo_level_3 = 2812

In [28]:
buildings_geo_3_2812 = train_values[train_values['geo_level_3_id']==2812]
buildings_geo_3_2812.head()

building_id  ...  has_secondary_use_other
1            28830  ...                        0
17093       265547  ...                        0
34007       674412  ...                        0
51151       988427  ...                        0
52872       111255  ...                        0

[5 rows x 39 columns]

In [32]:
buildings_geo_3_conditions = buildings_geo_3_2812.loc[(buildings_geo_3['position']=='s') & (buildings_geo_3['height_percentage']==7) & (buildings_geo_3['age']==10), :]
buildings_geo_3_conditions.head(5)

building_id  ...  has_secondary_use_other
1             28830  ...                        0
205258       938572  ...                        0

[2 rows x 39 columns]

Vemos que hay dos registros identicos, con la diferencia que tienen un building_id distinto. El significado de esto puede ser lo mencionado anteriormente, que simplemente al ser residenciales y están ubicados en el mismo geo_level_3, son edificios que se contruyeron de la misma forma, y con el mismo fin. Borrarlos sería un error, ya que perderíamos un 10% de la información, siendo diferentes registros.

Otro ejemplo puede ser el siguiente, donde el *building_id=590882* es igual a *building_id=1006803*

In [38]:
building_duplicated_2 = train_values.loc[(train_values['position']=='s') & (train_values['height_percentage']==5) & (train_values['age']==10) & (train_values['geo_level_3_id']==10694) & (train_values['other_floor_type']=='x'), :]
building_duplicated_2

building_id  ...  has_secondary_use_other
3            590882  ...                        0
138406       614984  ...                        0
142884      1006803  ...                        0

[3 rows x 39 columns]